docs/decisions/00TBD-function-calling-reliability.md

```diff
@@ -0,0 +1,246 @@
+- It can't address hallucinated function names. For instance, if the AI model generates the function FQN as `MyPlugin_my_func` instead of `MyPlugin_my_function`.
+
+Possible implementations:
+```csharp
+// Either at the operation level
+FunctionChoiceBehaviorOptions options = new new()
+{
+    FqnSeparator = "_"
+};
+
+var settings = new AzureOpenAIPromptExecutionSettings() { FunctionChoiceBehavior = FunctionChoiceBehavior.Auto(options) };
+
+var result = await this._chatCompletionService.GetChatMessageContentAsync(chatHistory, settings, this._kernel);
+
+// Or at the AI connector configuration level
+IKernelBuilder builder = Kernel.CreateBuilder();
+builder.AddOpenAIChatCompletion("<model-id>", "<api-key>", functionNamePolicy: FunctionNamePolicy.Custom("_"));
+```
+ 
+### Option 3: Custom FQN Parser
+
+This option proposes a custom, external FQN parser that can split function FQN into plugin name and function name. The parser will accepts the function FQN called by the AI model 
+and returns both the plugin name and function name. To achieve this, the parser will attempt to parse the FQN using various separator characters:
+```csharp
+static (string? PluginName, string FunctionName) ParseFunctionFqn(ParseFunctionFqnContext context)
+{
+    static (string? PluginName, string FunctionName)? Parse(ParseFunctionFqnContext context, char separator)
+    {
+        string? pluginName = null;
+        string functionName = context.FunctionFqn;
+
+        int separatorPos = context.FunctionFqn.IndexOf(separator, StringComparison.Ordinal);
+        if (separatorPos >= 0)
+        {
+            pluginName = context.FunctionFqn.AsSpan(0, separatorPos).Trim().ToString();
+            functionName = context.FunctionFqn.AsSpan(separatorPos + 1).Trim().ToString();
+        }
+
+        // Check if the function registered in the kernel
+        if (context.Kernel is { } kernel && kernel.Plugins.TryGetFunction(pluginName, functionName, out _))
+        {
+            return (pluginName, functionName);
+        }
+
+        return null;
+    }
+
+    // Try to use use hyphen, dot, and underscore sequentially as separators.
+    var result = Parse(context, '-') ??
+                    Parse(context, '.') ??
+                    Parse(context, '_');
+
+    if (result is not null)
+    {
+        return result.Value;
+    }
+
+    // If no separator is found, return the function name as is allowing AI connector to apply default behavior.
+    return (null, context.FunctionFqn);
+}
+```
+
+Pros:
+- It will mitigate but not reduce or completely eliminate function separator hallucinations by applying a custom heuristic specific to the AI model to parse the function FQN.
+- It can be easily implemented in SK AI connectors.
+
+Cons:
+- It can't be implemented in M.E.AI SK adapters because M.E.AI does not have the concept of plugins.
```

Comment:
Sure it could. It would be provided to the adapter, which creates M.E.AI function names from SK plugin/function names. It would just need to be passed through to the various places it was needed, e.g. by putting it on PromptExecutionSettings or FunctionChoiceBehavior, which by the way already has exposed separator information (https://github.com/microsoft/semantic-kernel/blob/7a61c9caff11f73bed93687144de3afda85fd39d/dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/FunctionChoiceBehavior.cs#L20) that would just be made settable / extensible.